File: mvp/BoardOutlook/BoardOutlook/Data/WebSocketMessage.swift
//
//  WebSocketMessage.swift
//  BoardOutlook
//
//  Created by lu on 18/4/2025.
//

import Foundation

enum WebSocketMessageType: String, Codable {
    case isUserReadyRequest = "IsUserReadyRequest"
    case isUserReadyResponse = "IsUserReadyResponse"
    case userInterviewAnswer = "UserInterviewAnswer"
}

struct UserInterviewAnswer: Codable {
    var userId: String
    var interviewId: String?
    var audioBase64: String?
    var isFirstAudioChunk: Bool
    var isLastAudioChunk: Bool
}

struct IsUserReadyResponse: Codable {
    var userId: String
    var isUserReady: Bool
    var transcript: String
}

enum WebSocketMessageData {
    case isUserReadyResponse(IsUserReadyResponse)
    case userInterviewAnswer(UserInterviewAnswer)
}

struct WebSocketMessage: Codable {
    let messageType: WebSocketMessageType
    let data: WebSocketMessageData

    enum CodingKeys: String, CodingKey {
        case messageType
        case data
    }

    init(messageType: WebSocketMessageType, data: WebSocketMessageData) {
        self.messageType = messageType
        self.data = data
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        messageType = try container.decode(WebSocketMessageType.self, forKey: .messageType)

        switch messageType {
        case .isUserReadyRequest, .userInterviewAnswer:
            let value = try container.decode(UserInterviewAnswer.self, forKey: .data)
            data = .userInterviewAnswer(value)
        case .isUserReadyResponse:
            let value = try container.decode(IsUserReadyResponse.self, forKey: .data)
            data = .isUserReadyResponse(value)
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(messageType, forKey: .messageType)

        switch data {
        case .isUserReadyResponse(let value):
            try container.encode(value, forKey: .data)
        case .userInterviewAnswer(let value):
            try container.encode(value, forKey: .data)
        }
    }
}

File: mvp/BoardOutlookApi/models/web_socket_message.py
from models.camel_model import CamelModel
from typing import Union, Optional
from enum import Enum
from pydantic import root_validator

class WebSocketMessageType(str, Enum):
    IS_USER_READY_REQUEST = "IsUserReadyRequest"
    IS_USER_READY_RESPONSE = "IsUserReadyResponse"
    USER_INTERVIEW_ANSWER = "UserInterviewAnswer"

class UserInterviewAnswer(CamelModel):
    user_id: str
    interview_id: Optional[str] = None
    audio_base64: Optional[str] = None
    is_first_audio_chunk: bool = True
    is_last_audio_chunk: bool = True

class IsUserReadyResponse(CamelModel):
    user_id: str
    is_user_ready: bool
    transcript: str

class WebSocketMessage(CamelModel):
    message_type: WebSocketMessageType
    data: Union[UserInterviewAnswer, IsUserReadyResponse]

    @root_validator(pre=True)
    def parse_data(cls, values):
        msg_type = values.get("messageType")  # camelCase from incoming message
        data = values.get("data")

        if not isinstance(data, dict):
            return values

        if msg_type == WebSocketMessageType.IS_USER_READY_RESPONSE:
            values["data"] = IsUserReadyResponse(**data)
        elif msg_type in {
            WebSocketMessageType.IS_USER_READY_REQUEST,
            WebSocketMessageType.USER_INTERVIEW_ANSWER,
        }:
            values["data"] = UserInterviewAnswer(**data)
        else:
            raise ValueError(f"Unknown message_type: {msg_type}")

        return values
